### ISSUE TITLE
Is it possible to do appDraw with IO?

### ISSUE THREAD

**Issue by anton-k** [jtdaugherty/brick#328]
Hi! 

My case is TUI-client with calls to API. It involves update of the data over API calls to server and display of the data alongside with some interaction with server. So I have calls that run in background and update visuals, and some calls that proposes changes to the remote server. 

I have to update stats periodically, it can be done with custom events in the brick.
But the problem lies in the fact that API calls take some time 0.5 to 1 seconds and during that time
the whole brick app is non-responsive fr interaction with user. 

I could solve this by running process in separate thread that queries data over API and saves it to TVars or MVars,
but unfortunately due to pure nature of appDraw I can not use those data directly.

One possiible workaround would be to have TVar's inside the state and in the tick call inside the brick
I do not perform API request but write data from TVar to internal brick state since liftIO is possible inside the handler.

but it means to have some boilerplate although it solves the problem, but if appDraw could let use IO
solution would be more strightforward.

**Comment by jtdaugherty** [comment]
Hi @anton-k, thanks for posting this.

> But the problem lies in the fact that API calls take some time 0.5 to 1 seconds and during that time
> the whole brick app is non-responsive fr interaction with user.
> I could solve this by running process in separate thread that queries data over API and saves it to TVars or MVars,
> but unfortunately due to pure nature of appDraw I can not use those data directly.

The way I suggest doing this is to use channels (i.e. `TChan`) rather than mutable variables (`TVar`) and have a design like this:

1. Spawn a background thread that waits on a channel for messages telling it what to do in interacting with the server API.
2. Have that background thread send messages back to the brick main event loop using a custom event.
3. Handle those custom events to update your application state.
4. Draw from your application state only.

The issue with doing what you ask is that not only is Brick designed specifically to prevent you from doing impure things in the drawing pass, even if it were possible somehow to do IO in the rendering step, that would create a whole new problem that would be very difficult to solve, which is that side effects triggered by drawing some part of the UI would need to potentially affect other parts of the UI, but if the parts are drawn in the wrong order, you'd have an inconsistent UI.

The whole point of Brick is to actually *help* lead to a cleaner, more maintainable application design with a *separation of concerns*. By concentrating where you handle events, where you keep state, and how you manage changes in state, the application design will end up much cleaner in the long run (and maybe even the short run).

I recognize that this might not be how you are used to writing applications, and if so, using Brick might require you to rethink your approach. In any case, I'm happy to answer more questions about the approach I described above; what you want to do is absolutely possible to do, but just not using the approach you are asking for. We use a version of the approach that I described in [the Matterhorn project](https://github.com/matterhorn-chat/matterhorn) where it has scaled extremely well despite enormous complexity in the server API and UI of the application.

(Although I'm declining the original request to do IO in drawing, I'm leaving this open so we can continue the discussion.)
